linera-core/src/client/mod.rs

```diff
@@ -1131,98 +1137,140 @@ where
-                committees.extend(new_committees);
-                max_epoch = max_epoch.max(new_max_epoch);
-            }
-            if block.epoch > max_epoch {
-                // We don't accept a certificate from a committee in the future.
-                warn!(
-                    "Postponing received certificate from {:.8} at height {} from future epoch {}",
-                    entry.chain_id, entry.height, block.epoch
-                );
-                // Stop the synchronization here. Do not increment the tracker further so
-                // that this certificate can still be downloaded later, once our committee
-                // is updated.
-                break;
-            }
-            match committees.get(&block.epoch) {
-                Some(committee) => {
-                    // This epoch is recognized by our chain. Let's verify the
-                    // certificate.
-                    certificate.check(committee)?;
-                    certificates.push(certificate);
-                    new_tracker += 1;
+
+            match self
+                .handle_certificate(
+                    admin_id,
+                    max_epoch,
+                    &mut committees,
+                    certificate,
+                    &remote_node,
+                )
+                .await?
+            {
+                HandleCertificateResult::FutureEpoch(certificate, new_max_epoch) => {
+                    max_epoch = new_max_epoch;
+                    warn!("Postponing received certificate from {:.8} at height {} from future epoch {}",
+                          entry.chain_id, entry.height, certificate.value().epoch());
+                    // Stop the synchronization here. Do not increment the tracker further so
+                    // that this certificate can still be downloaded later, once our committee
+                    // is updated.
+                    break;
                 }
-                None => {
+                HandleCertificateResult::OldEpoch(certificate) => {
                     // This epoch is not recognized any more. Let's skip the certificate.
                     // If a higher block with a recognized epoch comes up later from the
                     // same chain, the call to `receive_certificate` below will download
                     // the skipped certificate again.
                     warn!(
                         "Skipping received certificate from past epoch {:?}",
-                        block.epoch
+                        certificate.value().epoch()
                     );
                     new_tracker += 1;
                 }
+                HandleCertificateResult::New(certificate) => {
+                    certificates.push(certificate);
+                    new_tracker += 1;
+                }
             }
         }
         Ok((remote_node.name, new_tracker, certificates))
     }
 
+    async fn handle_certificate(
```

Comment:
Maybe `check_received_certificate`?

`handle_certificate` elsewhere always means that we update our state based on it, e.g. execute the block, update the chain state, and handle cross-chain messages.

Reply:
Done.